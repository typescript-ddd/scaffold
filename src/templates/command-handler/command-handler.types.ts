import {
  TemplateBooleanOption,
  TemplateEnumOption,
  TemplatePropertiesOption,
  TemplatePropertyValue,
  TemplateStringOption,
} from "../shared/template";

export type CommandHandlerTemplateValues = {
  entityName: string;
  actionName: string;
  actor: string;
  returnType: string;
  commandProperties: TemplatePropertyValue[];
  returnsView: boolean;
};

export type CommandHandlerTemplateOptions = {
  entityName: TemplateStringOption;
  actionName: TemplateEnumOption;
  actor: TemplateStringOption;
  returnType: TemplateStringOption;
  commandProperties: TemplatePropertiesOption;
  returnsView: TemplateBooleanOption;
};
