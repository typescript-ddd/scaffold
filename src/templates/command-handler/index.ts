export * from "./command-handler.generator";
export * from "./command-handler.template";
export * from "./command-handler.types";
