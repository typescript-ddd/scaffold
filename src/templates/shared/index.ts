export * from "./context-builder";
export * from "./generate-context";
export * from "./template";