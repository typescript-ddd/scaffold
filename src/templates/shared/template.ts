import { GenerateContext } from "./generate-context";

export type TemplateEnum = Record<string, string>;
export type TemplatePropertyValueType =
  | "string"
  | "number"
  | "boolean"
  | "date";

export type TemplateEnumOption = {
  name: string;
  label: string;
  description: string;
  enum: TemplateEnum;
  optionType: "enum";
};

export type TemplateStringOption = {
  name: string;
  label: string;
  description: string;
  optionType: "string";
};

export type TemplateTokenStringOption = {
  name: string;
  label: string;
  description: string;
  optionType: "tokenString";
};

export type TemplateBooleanOption = {
  name: string;
  label: string;
  description: string;
  optionType: "boolean";
};

export type TemplateNumberOption = {
  name: string;
  label: string;
  description: string;
  optionType: "number";
};

export type TemplateDateOption = {
  name: string;
  label: string;
  description: string;
  optionType: "date";
};

export type TemplatePropertyOption = {
  name: string;
  valueType: TemplatePropertyValueType;
  nullable?: boolean;
  prop?: string;
};

export type TemplatePropertyValue = {
  name: string;
  valueType: string;
  nullable?: boolean;
  prop?: string;
};

export type TemplatePropertiesOption = {
  name: string;
  label: string;
  description: string;
  optionType: "properties";
};

export type TemplateOption =
  | TemplateEnumOption
  | TemplateStringOption
  | TemplateBooleanOption
  | TemplateNumberOption
  | TemplateDateOption
  | TemplatePropertiesOption
  | TemplateTokenStringOption;

export type Maybe<T> = T | null | undefined;

export interface Template<
  TValues extends { [K in keyof TOptions]: any},
  TOptions extends { [key: string]: TemplateOption }
> {
  name: string;
  description: string;
  options: TOptions;
  defaultValues: TValues;
  generate(values: TValues, context: GenerateContext): string;
}