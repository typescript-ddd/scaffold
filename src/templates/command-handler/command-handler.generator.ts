import { Project, Scope } from "ts-morph";
import { strings } from "../../utils";
import { GenerateContext } from "../shared";
import { CommandHandlerTemplateValues } from "./command-handler.types";

const actorMap: Record<string, string> = {
  create: "creator",
  update: "updater",
  delete: "deleter",
};

export const generateCommandHandler = (
  values: CommandHandlerTemplateValues,
  context: GenerateContext
) => {
  const {
    entityName,
    actionName,
    actor,
    commandProperties,
    returnType,
    returnsView = false,
  } = values;
  const entityType = strings.capitalize(entityName);
  const actionType = strings.capitalize(actionName);
  const actorType = strings.capitalize(actor);
  const commandType = `${actionType}${entityType}Command`;
  const actorName = strings.lower(actorMap[actionName]);

  const className = `${actionType}${entityType}Handler`;

  const project = new Project();
  const sourceFile = project.createSourceFile(
    `${context.fileName}.ts`,
    `
export class ${className} implements CommandHandler<${commandType}> {}    
`
  );

  context.addImportDeclaration({
    moduleSpecifier: `@typescript-ddd/core`,
    namedImports: ["CommandHandler"],
  });

  context.addImportDeclaration({
    moduleSpecifier: context.resolveDir("application", "create"),
    namedImports: [commandType, actorType],
  });

  context.writeSourceFileImports(sourceFile);

  const classDeclaration = sourceFile.getClass(className);
  classDeclaration.addJsDoc({
    description: `Represents ${strings.describe(className)}.`,
  });

  const constructorDeclaration = classDeclaration.addConstructor({
    parameters: [
      {
        name: actorName,
        type: actorType,
        isReadonly: true,
        scope: Scope.Private,
      },
    ],
  });
  constructorDeclaration.addJsDoc({
    description: `Initializes a new instance of ${strings.describe(
      className
    )}.`,
    tags: [
      {
        tagName: "param",
        text: `{${actorType}} ${actorName} - The ${actorName} to handle the command.`,
      },
    ],
  });

  classDeclaration
    .addMethod({
      name: "handle",
      isAsync: true,
      returnType: `Promise<${returnType}>`,
      parameters: [{ name: "command", type: commandType }],
    })
    .setBodyText((writer) => {
      const isArray = returnType.endsWith("[]");
      const returnName = isArray
        ? strings.plural(strings.lower(entityName))
        : strings.lower(entityName);

      writer.writeLine(
        `const ${returnName} = await this.${actorName}.${strings.lower(
          actionName
        )}({`
      );
      commandProperties.forEach((property, index) => {
        writer.writeLine(
          `  ${property.name}: command.${property.name}${
            index === commandProperties.length - 1 ? "" : ","
          }`
        );
      });
      writer.writeLine(`});`);

      if (!returnsView) {
        writer.writeLine(`return ${returnName};`);
      } else {
        const viewCreator = returnType.endsWith("[]")
          ? returnType.slice(0, -2)
          : returnType;
        writer.writeLine(
          `return ${viewCreator}.${
            isArray ? "createMany" : "create"
          }(${returnName});`
        );
      }
    })
    .addJsDoc({
      description: `Handles ${strings.describe(commandType)}.`,
      tags: [
        {
          tagName: "param",
          text: `{${commandType}} command - The command to handle.`,
        },
        {
          tagName: "returns",
          text: `{Promise<${returnType}>} - ${strings.capitalize(
            strings.describe(returnType)
          )}.`,
        },
      ],
    });

  let output = sourceFile.getFullText();

  return output;
};
