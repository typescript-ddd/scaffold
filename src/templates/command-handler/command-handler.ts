import { Project, Scope } from "ts-morph";
import { strings } from "../../utils";
import { GenerateContext } from "../shared";

const actorMap: Record<string, string> = {
  create: "creator",
  update: "updater",
  delete: "deleter",
}

export type GenerateCommandHandlerOptions = {
  entityName: string;
  actionName: "create" | "update" | "delete";
  actor: string;
  returnType: string;
  commandProperties: { name: string; type: string }[];
};

export const generateCommandHandler = (
  options: GenerateCommandHandlerOptions,
  context: GenerateContext
) => {
  const { entityName, actionName, actor, commandProperties, returnType } =
    options;
  const entityType = strings.capitalize(entityName);
  const actionType = strings.capitalize(actionName);
  const actorType = strings.capitalize(actor);
  const commandType = `${actionType}${entityType}Command`;
  const actorName = strings.lower(actorMap[actionName]);

  const className = `${actionType}${entityType}Handler`;

  const project = new Project();
  const sourceFile = project.createSourceFile(
    `${context.fileName}.ts`,
    `
export class ${className} implements CommandHandler<${commandType}> {}    
`
  );

  context.addImportDeclaration({
    moduleSpecifier: `@typescript-ddd/core`,
    namedImports: ["CommandHandler"],
  });

  context.addImportDeclaration({
    moduleSpecifier: context.resolveDir("application", "create"),
    namedImports: [commandType, actorType],
  });

  context.writeSourceFileImports(sourceFile);

  const classDeclaration = sourceFile.getClass(className);
  classDeclaration.addJsDoc({
    description: `Represents ${strings.describe(className)}.`,
  });

  const constructorDeclaration = classDeclaration.addConstructor({
    parameters: [{ name: actorName, type: actorType, isReadonly: true, scope: Scope.Private }],
  });
  constructorDeclaration.addJsDoc({
    description: `Initializes a new instance of ${strings.describe(
      className
    )}.`,
    tags: [
      {
        tagName: "param",
        text: `{${actorType}} ${actorName} - The ${actorName} to handle the command.`,
      },
    ],
  });

  classDeclaration
    .addMethod({
      name: "handle",
      isAsync: true,
      returnType: `Promise<${returnType}>`,
      parameters: [{ name: "command", type: commandType }],
    })
    .setBodyText((writer) => {
      writer.writeLine(
        `const ${strings.lower(entityName)} = await this.${actorName}.${strings.lower(
          actionName
        )}({`
      );
      commandProperties.forEach((property, index) => {
        writer.writeLine(
          `  ${property.name}: command.${property.name}${
            index === commandProperties.length - 1 ? "" : ","
          }`
        );
      });
      writer.writeLine(`});`);
      writer.writeLine(
        `return ${returnType}.${strings.lower(actionName)}(${strings.lower(
          entityName
        )});`
      );
    })
    .addJsDoc({
      description: `Handles ${strings.describe(commandType)}.`,
      tags: [
        {
          tagName: "param",
          text: `{${commandType}} command - The command to handle.`,
        },
        {
          tagName: "returns",
          text: `{Promise<${returnType}>} - ${strings.capitalize(strings.describe(returnType))}.`,
        },
      ],
    })

  let output = sourceFile.getFullText();

  return output;
};
