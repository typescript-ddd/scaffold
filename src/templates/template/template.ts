import { Project, Scope } from "ts-morph";
import { strings } from "../../utils";
import { GenerateContext } from "../shared";

export type GenerateTemplateOptions = {
  name: string;
};

export const generateTemplate = (
  options: GenerateTemplateOptions,
  context: GenerateContext
) => {
  const { name } = options;
  const className = strings.capitalize(name);

  const project = new Project();
  const sourceFile = project.createSourceFile(
    `${context.fileName}.ts`,
    `
export class ${className} {}    
`
  );

  context.addImportDeclaration({
    moduleSpecifier: `@typescript-ddd/core`,
    namedImports: ["Entity"],
  });

  context.writeSourceFileImports(sourceFile);

  const classDeclaration = sourceFile.getClass(className);
  classDeclaration.addJsDoc({
    description: `${strings.describe(name)}.`,
  });

  const constructorDeclaration = classDeclaration.addConstructor({
    scope: Scope.Private,
    parameters: [{ name: "value", type: "string" }],
  });
  constructorDeclaration.setBodyText(`super(value);`);
  constructorDeclaration.addJsDoc({
    description: `Initializes a new instance of ${strings.describe(
      name
    )} identifier.`,
    tags: [{ tagName: "param", text: "value The value of the identifier." }],
  });

  let output = sourceFile.getFullText();

  return output;
};
