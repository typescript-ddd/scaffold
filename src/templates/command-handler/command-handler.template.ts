import { GenerateContext } from "../shared";
import { Template } from "../shared/template";
import { generateCommandHandler } from "./command-handler.generator";
import {
  CommandHandlerTemplateOptions,
  CommandHandlerTemplateValues,
} from "./command-handler.types";

export class CommandHandlerTemplate
  implements
    Template<CommandHandlerTemplateValues, CommandHandlerTemplateOptions>
{
  readonly name = "Command";
  readonly description = "Generates a command";
  readonly options: CommandHandlerTemplateOptions = {
    actionName: {
      name: "actionName",
      label: "Action Name",
      description: "The name of the action",
      optionType: "enum",
      enum: { create: "Create", update: "Update", delete: "Delete" },
    },
    entityName: {
      name: "entityName",
      label: "Entity Name",
      description: "The name of the entity",
      optionType: "string",
    },
    actor: {
      name: "actor",
      label: "Actor",
      description: "The actor of the command",
      optionType: "string",
    },
    returnType: {
      name: "returnType",
      label: "Return Type",
      description: "The return type of the command",
      optionType: "string",
    },
    commandProperties: {
      name: "commandProperties",
      label: "Properties",
      description: "The properties of the command",
      optionType: "properties",
    },
    returnsView: {
      name: "returnsView",
      label: "Returns View",
      description: "Whether the command returns a view",
      optionType: "boolean",
    },
  };
  readonly defaultValues = {
    actionName: "create",
    entityName: "User",
    actor: "UserCreator",
    returnType: "UserView",
    commandProperties: [{ name: "name", valueType: "string" }],
    returnsView: true,
  };
  generate(values: CommandHandlerTemplateValues, context: GenerateContext): string {
    return generateCommandHandler(values, context);
  }
}
