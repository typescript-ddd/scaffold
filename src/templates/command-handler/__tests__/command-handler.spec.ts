import { ContextBuilder, GenerateContext } from "../../shared";
import { CommandHandlerTemplate } from "../command-handler.template";

describe("CommandHandler", () => {
    let context: GenerateContext;
    let template: CommandHandlerTemplate;

    beforeEach(() => {
        context = ContextBuilder.build("@src");
        template = new CommandHandlerTemplate();
    })

    it("Should render the command handler template", () => {
        
        context.addImportDeclaration({
            moduleSpecifier: context.resolveDir("domain", "view"),
            namedImports: ["UserView"],
        });

        const output = template.generate(
            {
                entityName: "User",
                actionName: "create",
                actor: "UserCreator",
                returnType: "UserView",
                returnsView: true,
                commandProperties: [
                    { name: "name", valueType: "string" },
                ]
            },
            context
        );

        expect(output).toMatchSnapshot();
    });    
})