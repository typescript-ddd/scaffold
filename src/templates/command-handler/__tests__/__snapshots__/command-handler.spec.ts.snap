// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CommandHandler Should render the command handler template 1`] = `
"import { CommandHandler } from "@typescript-ddd/core";
import { CreateUserCommand, UserCreator } from "@src/application/create";
import { UserView } from "@src/domain/view";

/** Represents a create user handler. */
export class CreateUserHandler implements CommandHandler<CreateUserCommand> {
    /**
     * Initializes a new instance of a create user handler.
     * @param {UserCreator} creator - The creator to handle the command.
     */
    constructor(private readonly creator: UserCreator) {
    }

    /**
     * Handles a create user command.
     * @param {CreateUserCommand} command - The command to handle.
     * @returns {Promise<UserView>} - A user view.
     */
    async handle(command: CreateUserCommand): Promise<UserView> {
        const user = await this.creator.create({
          name: command.name
        });
        return UserView.create(user);
    }
}    
"
`;
